### ISSUE TITLE
Cannot map Object anotated with lombok @Builder when two sources present

### ISSUE THREAD

**Issue by hurynovich** [mapstruct/mapstruct#3353]
### Expected behavior

When `target` object has `@Builder` annotation from lombok and this target has nested object which populated from two or more sources then compilation fails with following message:
```
java: No read accessor found for property "customer" in target type.
```

### Actual behavior

Here I created sample project which reproduces the problem when project is build with `mvn clean install`:
https://github.com/hurynovich/mastruct-builder-bug/tree/main
Build fails.

### Steps to reproduce the problem

Here I created PR in my sample project where I removed `@Builder` annotation:
https://github.com/hurynovich/mastruct-builder-bug/pull/1/files
Build successful.

### MapStruct Version

1.5.5.Final

**Comment by hurynovich** [comment]
I also found similar problem which was closed due to lack of feedback:
https://github.com/mapstruct/mapstruct/issues/2226

**Comment by thunderhook** [comment]
This is not a problem with lombok, but with the builder pattern in combinations like these.

To save the others from having to check out the code, here the relevant code parts:

```java
// Source

@Data
public class OrderDto {
    private String customerId;
    private PersonDto customerData;
}

@Data
public class PersonDto {
    private String name;
    private String surname;
}

// Target

@Data
@Builder
public class Order {
    private Person customer;
}

@Data
public class Person {
    private String id;
    private String name;
    private String surname;
}

@Mapper
public interface DtoMapper {

    @Mapping(target = "customer.id", source = "customerId")
    @Mapping(target = "customer", source = "orderDto.customerData")
    Order mapOrder(OrderDto orderDto);

}
```

Let's see what is generated without the `@Builder` annotation:

```java
public class DtoMapperImpl implements DtoMapper {

    @Override
    public Order mapOrder(OrderDto orderDto) {
        if ( orderDto == null ) {
            return null;
        }

        Order order = new Order();

        if ( orderDto.getCustomerData() != null ) {
            if ( order.getCustomer() == null ) { // <-- important part!
                order.setCustomer( new Person() );
            }
            personDtoToPerson( orderDto.getCustomerData(), order.getCustomer() );
        }
        if ( order.getCustomer() == null ) { // <-- important part!
            order.setCustomer( new Person() );
        }
        orderDtoToPerson( orderDto, order.getCustomer() );

        return order;
    }

    protected void personDtoToPerson(PersonDto personDto, Person mappingTarget) {
        if ( personDto == null ) {
            return;
        }

        mappingTarget.setName( personDto.getName() );
        mappingTarget.setSurname( personDto.getSurname() );
    }

    protected void orderDtoToPerson(OrderDto orderDto, Person mappingTarget) {
        if ( orderDto == null ) {
            return;
        }

        mappingTarget.setId( orderDto.getCustomerId() );
    }
}

```

In order for mapstruct to know whether the `Person` must be initialized or not, it checks the corresponding getter:
```java
if ( order.getCustomer() == null ) { // <-- important part!
    order.setCustomer( new Person() );
}
```

When you change it to a target with a builder pattern (wheter by hand or lombok, it doesn't matter), the code should like this:

```java
    public Order mapOrder(OrderDto orderDto) {
        if ( orderDto == null ) {
            return null;
        }

        Order.OrderBuilder builder = Order.builder();

        if ( orderDto.getCustomerData() != null ) {
            if ( builder.getCustomer() == null ) {
                builder.customer( new Person() );
            }
            personDtoToPerson( orderDto.getCustomerData(), builder.getCustomer() );
        }
        if ( builder.getCustomer() == null ) {
            builder.customer( new Person() );
        }
        orderDtoToPerson( orderDto, builder.getCustomer() );

        return builder.build();
    }
```

The check if a customer is already initialized is only possible with a getter of `customer` in the mapping target, which is a builder in that case.
This is why @filiphr asked about the getter in the comment https://github.com/mapstruct/mapstruct/issues/2226#issuecomment-706506724

So, in your case, you can either disable the builder for mapping. Or you can extend the builder with a getter like this:
```java
@Data
@Builder
public class Order {
    private Person customer;

    public static class OrderBuilder {
        public Person getCustomer() {
            return customer;
        }
    }

}
```

**Comment by thunderhook** [comment]
> Or you can extend the builder with a getter like this: ...

Sidenote: this only works when you have a mutable target class that can be modified with setters.
As soon as you use a immutable target (annotated with `@Value` and `@Builder`) this won't work.

There is another option for you that I stumbled accross.

As mentioned in https://github.com/mapstruct/mapstruct/issues/2661#issuecomment-981084182 and https://github.com/mapstruct/mapstruct/issues/2661#issuecomment-1278985283 you can "split" the mappings like this:

```java
@Mapper
public interface DtoMapper {

    @Mapping(target = "customer", source = ".")
    Order mapOrder(OrderDto order);

    @Mapping(target = "id", source = "customerId")
    @Mapping(target = ".", source = "customerData")
    Person mapPerson(OrderDto order);

}
```

This approach works for mutable or immutable targets, and it does not matter how many levels the nested properties are (imagine an `Address` entity inside Person that must be mapped from `OrderDto.addressId`).

@filiphr I don't know exactly if MapStruct can improve the support of nested properties from different levels/layers of source parameters. If so, I can try to help.

**Comment by thunderhook** [comment]
This might be solved as well with #2941 which is exactly about improving this kind of "inbalanced source parameters".

**Comment by filiphr** [comment]
Thanks a lot for checking out the repo and providing this detailed feedback @thunderhook.

Indeed for us to create the right data when using nested mappings we need to be able to create the intermediate objects. 

Thanks for pointing out to #2941 @thunderhook. Ideally we can resolve this through that one. However, the current solution is what @thunderhook shared in https://github.com/mapstruct/mapstruct/issues/3353#issuecomment-1699007863

I will only add one note for the example. Instead of using `source = "."` we suggest using the name of the parameter e.g. `source = "order"`. The use of `source = "."` isn't really documented and it isn't well tested I believe.

e.g.

```java
@Mapper
public interface DtoMapper {

    @Mapping(target = "customer", source = "order")
    Order mapOrder(OrderDto order);

    @Mapping(target = "id", source = "customerId")
    @Mapping(target = ".", source = "customerData")
    Person mapPerson(OrderDto order);

}
```

I am going to close this one since we have #2941 for improving the nested parameter mappings.
